File: leetcode/0053.MaximumSubarray/Solution.cpp
#include "../../include/global.hpp"
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    int maxSubArray(vector<int>& nums) {
        int thissum = 0;
        int maxsum = INT32_MIN;
        for (auto n : nums) {
            thissum = thissum + n;
            if (thissum < 0) {
                maxsum = max(maxsum, thissum);
                thissum = 0;
            } else
                maxsum = max(maxsum, thissum);
        }
        return maxsum;
    }
};

int main() {
    vector<vector<int>> samples = {
        {-2, 1, -3, 4, -1, 2, 1, -5, 4}, {0}, {-2}, {-1, -2}};
    Solution s;
    for (auto i : samples)
        cout << s.maxSubArray(i) << endl;
}

File: leetcode/0053.MaximumSubarray/Solution.go
package main

import (
	"fmt"
	"slices"
)

func maxSubArray(nums []int) int {
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		dp[i] = max(dp[i-1]+nums[i], nums[i])
	}
	return slices.Max(dp)
}

func main() {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	output := maxSubArray(nums)
	fmt.Println(output)
}
